Feature/jina lee11 store

jinaLee11 opened the pull request.

jinaLee11 committed:
feat: 가게 관련 패키지, 클래스 생성

jinaLee11 committed:
Merge remote-tracking branch 'origin' into feature/jinaLee11-store

jinaLee11 committed:
feat: 가게 생성 기능 구현

jinaLee11 committed:
feat: 가게 다건 조회

jinaLee11 committed:
feat: 가게 단건 조회

jinaLee11 committed:
feat: 가게 수정 및 삭제

- 아직 postman 동작 미확인
- 추후 메뉴리스트 추가 필요

jinaLee11 committed:
feat: 가게 CRUD 구현 완료

- 추후 메뉴리스트 추가 필요

jinaLee11 committed:
feat: 메뉴 CRUD 파일 준비, 패키지 변경 pull 전 commit

jinaLee11 committed:
Merge remote-tracking branch 'origin' into feature/jinaLee11-store

jinaLee11 committed:
feat: 메뉴 생성 구현

jinaLee11 committed:
feat: 메뉴 수정, 삭제 구현

jinaLee11 committed:
feat: 메뉴 조회 구현

- 가게 단일조회시 메뉴 리스트까지 출력하는 기능 구현
- Postman으로 확인 시 오류 발생하는 부분 있음. 수정 필요한 상태

jinaLee11 committed:
fix: 메뉴 삭제 반영 안되는 부분 수정

jarvisRyu commented on src/main/java/org/example/delivery/domain/menu/service/MenuService.java:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        // 3. 메뉴 리퀘스트 받아온걸로 수정
+        if (requestDto.getName() != null) {
+            menu.updateName(requestDto.getName());
+        }
+
+        if (requestDto.getPrice() != null) {
+            menu.updatePrice(requestDto.getPrice());
+        }
+
+        if (requestDto.getContent() != null) {
+            menu.updateContent(requestDto.getContent());
+        }
```
이부분은 RequestDto에서 @Valid 예외처리를 사용하면 더 좋을듯 합니다.

crocusia replied:
비밀번호 같은 비즈니스 로직이 필요하지 않은 간단한 정보를 수정할 때는
PATCH가 아닌 PUT을 사용하는 방식이 선호된다고 합니다.
PUT을 사용할 경우, 수정하지 않을 정보는 프론트에서 기존의 값을 그대로 입력해주는 방식이기 때문에
서비스 계층에서 null을 검사하지 않아도 되고, DTO에서 유효성 검증을 할 수 있다는 장점이 있습니다 :)
ex) 수정 가능한 정보 : 이름, 가격, 내용
      수정하고 싶은 정보 : 가격
      request
      {
        "이름" : "기존 그대로 - 프론트가 넣어줌",
        "가격" : 변경할 가격,
        "이름" : "기존 그대로 - 프론트가 넣어줌"
      }
한번 고려해보시는 것도 좋을 것 같아요

le951 commented on src/main/java/org/example/delivery/domain/store/entity/Store.java:
```diff
@@ -20,6 +25,8 @@ public class Store {
+
     private LocalTime openingTime;
 
     private LocalTime closing_time;
```
closingTime이 되어야 할 것 같습니다.
dev도 똑같은걸 보니 제가 init 과정에서 오타낸 것 같습니다. 미안합니다.

그래도 수정을 하긴 해야하고
제가 건들기보다는 해당 domain 작업중이신 분이 하시는 편이
혹시 모를 꼬임을 방지할 수 있을 것 같습니다.

le951 replied:
IntelliJ Refactor 기능으로 전체 수정 한 번 시원하게 때려주시면 감사하겠습니다

le951 replied:
수정 하시고 다시 push 하시면 아마 자동 반영 될겁니다.

le951 commented on src/main/java/org/example/delivery/domain/menu/dto/MenuRequestDto.java:
```diff
@@ -0,0 +1,20 @@
+@RequiredArgsConstructor
+public class MenuRequestDto {
+
+    @NotNull
```
NotNull이면 " " 이나 "     " 같은 공백만으로 구성되는 내용이 가능할겁니다.
NotBlank나 Pattern 어노테이션 사용하여 처리하는 것이 좋을 것 같습니다.

le951 commented on src/main/java/org/example/delivery/domain/store/dto/StoreRequestDto.java:
```diff
@@ -0,0 +1,25 @@
+    private final LocalTime openingTime; // 오픈시간, 포맷팅 고민
+
+    @NotNull
+    private final LocalTime closing_time; // 마감시간
```
여기도 마찬가지로 NotNull과 closing_time 수정이 필요할 것 같습니다.

crocusia replied:
@JsonFormat(pattern = "HH:mm") 이런 식의 어노테이션으로 정해진 형식을 받는 법도 있습니다!

crocusia commented on src/main/java/org/example/delivery/domain/menu/service/MenuService.java:
```diff
@@ -0,0 +1,99 @@
+            throw new IllegalArgumentException("접근 권한이 없습니다.");
+        }
+
+        Menu menu = new Menu(
```
팩토리 메서드 방식을 추천 드립니다!

le951 commented on src/main/java/org/example/delivery/domain/menu/service/MenuService.java:
```diff
@@ -0,0 +1,99 @@
+
+        // 2. request 체크해서 로그인한 사장님만 수정 가능
+        Long userId = (Long) request.getAttribute("userId"); // 로그인한 user id 가져옴
+        if (!menu.getStore().getOwner().getId().equals(userId)){ // 메뉴랑 매핑된 가게의 주인의 아이디 가져와서 로그인 아디랑 비교
```
이 체이닝 처리들은 구조상 여러 차례의 Query를 거칠 것 같습니다.

JPA의 기본은 FetchType.LAZY 입니다.
현재 Menu.store 는 @​ManyToOne 만 되어있습니다.
따라서 지연로딩 합니다.

1. (menu는 위에서 findById)
2. getStore 쿼리 요청 = Store 값 받음
3. getOwner 쿼리 요청 = User 값 받음
4. (이후 가져온 User 값과 비교. 메모리 내 접근)

따라서 두 번의 쿼리가 실행됩니다. N+1 같습니다.

비슷한 패턴이 여럿 있는 것 같습니다. 바로 수정되진 않을 것 같은데
당장 작동은 되니까 Issue로 달아두고, 일단 넘기는것도 괜찮아 보입니다.

le951 reviewed (changes requested):
쿼리 성능 문제는 당장 작동만 된다면
Issue로 달아두고 내일 수정하셔도 괜찮을 것 같습니다만

가능하시면 NotNull과 closing_time -> closingTime
이 두 개는 바로 수정해서 push 해주시면 좋겠습니다.

crocusia commented on src/main/java/org/example/delivery/domain/menu/controller/MenuController.java:
```diff
@@ -0,0 +1,54 @@
+    @PostMapping
+    public ResponseEntity<MenuResponseDto> createMenu(
+            @PathVariable Long storeId,
+            @RequestBody MenuRequestDto requestDto,
```
앞에 @Valid를 붙여주셔야 실제로 검증이 되는 걸로 알고 있습니다
@Valid  @RequestBody

jinaLee11 committed:
fix: notNull 변경, closing_time 변수명 수정

- @notNull -> @notBlank 로 수정
- @Valid 누락된 부분 수정
- closing_time -> closingTime 으로 수정

le951 committed:
style: closing_time -> closingTime

le951 reviewed (approved):
쿼리 성능 문제는 별도 issue 등록
다른 작업과의 병합을 위해 Merge
